### PR TITLE
Vp/min 222 get hypothetical rpc

### PR DESCRIPTION
**Pull request by dev-vadim**
**Description:**

Create RPC to be able to call get_hypothetical_account_liquidity from frontend

**Changes in storage or API:**

Add controller_hypotheticalAccountLiquidity to controller RPC

Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

**Checklist**

- [ ] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: runtime/src/tests/rpc_tests.rs [hunk 1]
```diff
@@ -319,6 +319,31 @@ fn test_user_balance_using_rpc() {
...
 }
 
+#[test]
+fn test_get_hypothetical_account_liquidity_rpc() {
+	ExtBuilder::default()
+		.pool_initial(DOT)
+		.pool_initial(ETH)
+		.build()
+		.execute_with(|| {
+			// Set price = 2.00 USD for all polls.
```
Comment (DenisRomanovsky) [inline runtime/src/tests/rpc_tests.rs:329]:
```suggestion
			// Set price = 2.00 USD for all pools.
```

Comment (dev-vadim) [reply runtime/src/tests/rpc_tests.rs]: Fixed

File: pallets/controller/rpc/src/lib.rs [hunk 2]
```diff
@@ -1,23 +1,35 @@
...
 	#[rpc(name = "controller_userBalanceInfo")]
 	fn get_user_balance(&self, account_id: AccountId, at: Option<BlockHash>) -> Result<Option<UserPoolBalanceData>>;
 
+	#[rpc(name = "controller_hypotheticalAccountLiquidity")]
+	fn get_hypothetical_account_liquidity(
+		&self,
+		account_id: AccountId,
+		underlying_to_borrow: CurrencyId,
+		redeem_amount: Balance,
+		borrow_amount: Balance,
```
Comment (lut1k) [inline pallets/controller/rpc/src/lib.rs:29]: Denis should write about this.

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -766,6 +767,34 @@ impl_runtime_apis! {
...
+				Ordering::Less => {
+					let diff = shortfall.checked_sub(excess)?;
+					let amount = Amount::try_from(diff).ok()?;
+					amount.checked_neg()?
+				},
+				_ => {
+					let diff = excess.checked_sub(shortfall)?;
+					Amount::try_from(diff).ok()?
+				}
+			};
```
Comment (lut1k) [inline runtime/src/lib.rs:793]: I think this is incorrect. You have to give to the Frontend (excess, shortfall).

File: pallets/controller/rpc/src/lib.rs (hunk 2 above)
Comment (dev-vadim) [reply pallets/controller/rpc/src/lib.rs]: Removed these parameters

File: pallets/controller/rpc/src/lib.rs [hunk 4]
```diff
@@ -1,23 +1,35 @@
...
 #[rpc]
-pub trait ControllerApi<BlockHash, AccountId> {
+pub trait ControllerApi<BlockHash, AccountId, Balance> {
 	#[rpc(name = "controller_liquidityPoolState")]
 	fn liquidity_pool_state(&self, pool_id: CurrencyId, at: Option<BlockHash>) -> Result<Option<PoolState>>;
 
 	#[rpc(name = "controller_userBalanceInfo")]
 	fn get_user_balance(&self, account_id: AccountId, at: Option<BlockHash>) -> Result<Option<UserPoolBalanceData>>;
 
+	#[rpc(name = "controller_hypotheticalAccountLiquidity")]
```
[the remark this answers is not in the document]
Comment (dev-vadim) [reply pallets/controller/rpc/src/lib.rs]: Done

File: runtime/src/lib.rs (hunk 3 above)
Comment (dev-vadim) [reply runtime/src/lib.rs]: Simplified logic according to provided info about get_hypothetical_account_liquidity() return values

**Review by lut1k — APPROVED** [review]

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -758,6 +758,19 @@ impl_runtime_apis! {
...
+			let res = match excess.cmp(&shortfall) {
+				Ordering::Less => {
+					let amount = Amount::try_from(shortfall).ok()?;
+					amount.checked_neg()?
+				},
+				_ => Amount::try_from(excess).ok()?
+			};
+
+			Some(HypotheticalLiquidityData{ liquidity: res })
+		}
```
Comment (lut1k) [inline runtime/src/lib.rs:772]: Played with the syntax :)
```
fn get_hypothetical_account_liquidity(account_id: AccountId) -> Option<HypotheticalLiquidityData> {
			let (excess, shortfall) = Controller::get_hypothetical_account_liquidity(&account_id, MNT, 0, 0).ok()?;

			Some(HypotheticalLiquidityData{ liquidity:
				if !excess.is_zero() {
				excess.try_into().ok()?
			} else {
				Amount::try_from(shortfall).map(|v| v.checked_neg()).ok()??
			} })
		}
```

**Review by kurotychb — APPROVED** [review]

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -758,6 +758,19 @@ impl_runtime_apis! {
...
+			let (excess, shortfall) = Controller::get_hypothetical_account_liquidity(&account_id, MNT, 0, 0).ok()?;
+			let res = match excess.cmp(&shortfall) {
+				Ordering::Less => {
+					let amount = Amount::try_from(shortfall).ok()?;
+					amount.checked_neg()?
+				},
+				_ => Amount::try_from(excess).ok()?
+			};
+
+			Some(HypotheticalLiquidityData{ liquidity: res })
```
Comment (Apollo50) [inline runtime/src/lib.rs:771]: I like you keep only one parameter in structure.

**Review by Apollo50 — APPROVED** [review]

**Review by DenisRomanovsky — APPROVED** [review]

**Review by DenisRomanovsky — APPROVED** [review]

**Pull request merged**